### DISCUSSION TITLE
Consider using Floating UI

### DISCUSSION THREAD

**Discussion by atomiks** [mantinedev/mantine#490] [Feedback]
Hi,

Amazing work on this library. I am one of the co-authors of Popper v2, and I created a new project called [Floating UI](https://www.floating-ui.com) as a next-generation evolution of it. You can read the [motivation for creating it here](https://www.floating-ui.com/docs/motivation).

In summary, it's designed to be an ideal solution for component libraries like this one because of its super low-level, unopinionated nature. It's smaller, faster, more predictable, and is tree-shakeable + strongly typed out of the box. I think you'll enjoy authoring any floating elements using it!

Cheers

**Comment by bajoski34** [comment]
👍

**Comment by rtivital** [comment]
Hi @atomiks, hope you would not mind several questions (sorry, I don't have time to go through all docs):
- Will we be able to have all of our current Popper component features (examples: [Select](https://github.com/mantinedev/mantine/blob/master/src/mantine-core/src/components/Select/SelectDropdown/SelectDropdown.tsx), [Popper](https://github.com/mantinedev/mantine/blob/master/src/mantine-core/src/components/Popper/Popper.tsx) itself)?
- Is it compatible with popper.js API? Will we have to rewrite Popper component logic entirely to migrate?
- Am I right that you have a built-in middleware similar to what we have in [Select component](https://github.com/mantinedev/mantine/blob/master/src/mantine-core/src/components/Select/SelectDropdown/SelectDropdown.tsx#L72-L86) to make dropdown same width as input?
- Do you have an estimation on how much bundle size it will save? Currently, [according to bundlephobia](https://bundlephobia.com/package/@mantine/core@3.1.7) popper is about 30kb.

@EmilMalanczak you've worked with popper more than me, so what do you think? Maybe you have some additional questions?

**Reply by atomiks** [reply]
- Everything you can do with Popper, you can do with Floating UI, since it's lower-level
- No it's not compatible, which is why it's probably better for a major version, or you can transform it to match the current one
- It doesn't right now, but can be added pretty easily as a custom middleware, might include it in core since it's a common use case
- Floating UI is less than half the size [according to Bundlephobia](https://bundlephobia.com/package/@floating-ui/react-dom@0.3.0), but it also supports tree-shaking, so in many cases it can be even less

**Comment by EmilMalanczak** [comment]
I've seen announcements about Floating UI on a daily.dev 😄 . I like the API it uses but never thought about using it here. Bundle size would shrink with such change for sure. It doesn't seem to be THAT much hard to migrate but it would be time-consumable.

- Is there something like the `enable` boolean key for turning on/off middleware? like this example we use
```tsx
{
  name: 'preventOverflow',
  enabled: false,
},
```

- According to this snippet:
```tsx
 const {
    x,
    y,
    reference,
    floating,
    middlewareData: {arrow: {x: arrowX, y: arrowY} = {}},
  } = useFloating({
    middleware: [arrow({element: arrowRef})],
  });
```
Any data we will pass to middlewares there will be later on available under the `middlewareData` object, right? 

- Are there any potential problem due to the `asynchronous` logic of the `detectOverflow` function?  

- That function is kinda a must-have (we had some problems with popper here) for us to make it work our Select components. 
> * Am I right that you have a built-in middleware similar to what we have in [Select component](https://github.com/mantinedev/mantine/blob/master/src/mantine-core/src/components/Select/SelectDropdown/SelectDropdown.tsx#L72-L86) to make dropdown the same width as input?


In summary, I believe that might be a good change - lib seems similar to popper v3 😄 It might be a good point for 3.5
What are your plans for higher-level components you mention on github? @atomiks I'm curious about it.

**Reply by atomiks** [reply]
- You control the `middleware` array, so you can do filtering on the array yourself to achieve the same behavior. Nothing is "enabled" by default like in Popper, you have to manually supply all the middleware you want, so you can just filter things or use if/ternary conditions where you call `computePosition` to avoid passing in a certain middleware
- Those are the options to configure the behavior. Data is different — you can read the [Middleware docs](https://www.floating-ui.com/docs/middleware) for more details
- The asynchronicity of Floating UI on the web is "fake" and it only uses a Promise microtask, so there are no issues. The only reason it's async is to support platforms other than the web.
- I'm going to modify the `size` middleware to enable this feature easily. It only takes one line actually, a single extra argument

Regarding the higher-level components, this one's tricky because there are so many different solutions for this. The low-level nature of positioning was hard but also easy at the same time because you don't have many opinions on anything. I have to keep exploring how it will work over the next months, but it will be kinda like Radix's philosophy but possibly more generic — unstyled, extensible with accessible defaults

**Comment by rtivital** [comment]
Thanks for the answers @atomiks, we will consider migrating.
